src/Compilers/Core/CodeAnalysisTest/Diagnostics/SystemLanguageAnalyzer.cs

```diff
@@ -128,4 +132,118 @@ void Report(CompilationAnalysisContext context, IFieldSymbol field, DiagnosticDe
             context.ReportDiagnostic(Diagnostic.Create(descriptor, field.Locations.FirstOrDefault()));
         }
     }
+
+    /// <summary>Analyzer used to identify local variables that could be declared Const.</summary>
+    public class LocalCouldBeConstAnalyzer : DiagnosticAnalyzer
+    {
+        private const string SystemCategory = "System";
+
+        public static readonly DiagnosticDescriptor LocalCouldBeConstDescriptor = new DiagnosticDescriptor(
+            "LocalCouldBeReadOnly",
+            "Local Could Be Const",
+            "Local variable is never modified and so could be const.",
+            SystemCategory,
+            DiagnosticSeverity.Warning,
+            isEnabledByDefault: true);
+
+        /// <summary>Gets the set of supported diagnostic descriptors from this analyzer.</summary>
+        public sealed override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics
+        {
+            get { return ImmutableArray.Create(LocalCouldBeConstDescriptor); }
+        }
+
+        public sealed override void Initialize(AnalysisContext context)
+        {
+            context.RegisterOperationBlockStartAction(
+                (operationBlockContext) =>
+                {
+                    IMethodSymbol containingMethod = operationBlockContext.OwningSymbol as IMethodSymbol;
+
+                    if (containingMethod != null)
+                    {
+                        HashSet<ILocalSymbol> mightBecomeConstLocals = new HashSet<ILocalSymbol>();
+                        HashSet<ILocalSymbol> assignedToLocals = new HashSet<ILocalSymbol>();
+
+                        operationBlockContext.RegisterOperationAction(
```

Comment:
The intent is that the analyzer work correctly regardless of action execution order. Looking at the code, there's a bug in that the symbol action should check that a symbol is not in assignedToLocals before putting it in mightBecomeConstLocals.

This is especially embarrassing in that the FieldCouldBeReadOnly analyzer does not have this bug, and I thought I had written this analyzer to be action order independent.

Thanks!